Fix/4556

zah opened the pull request:
This was quite tricky to get right.

Take a look at the test case in the last commit to see what kinds of new checks are performed.

Araq commented on compiler/semexprs.nim:
```diff
@@ -2035,25 +2035,238 @@ proc isTupleType(n: PNode): bool =
+    initNone     # None of the fields have been initialized
+    initConflict # Fields from different branches have been initialized
+
+proc mergeInitStatus(existing: var InitStatus, newStatus: InitStatus) =
```
This is now complex enough so that it should be put into a ``semobjconstr.nim`` include file.

zah replied:
Alright, but let's first merge the fix as it is. I'm already working on something else and juggling these branches takes away from my limited time.

Araq replied:
Oh it's not necessary to have a branch per bugfix. Just a single "zahary" branch that you request me to merge from time to time is enough.

Araq commented on compiler/semexprs.nim:
```diff
@@ -2035,25 +2035,238 @@ proc isTupleType(n: PNode): bool =
+      if field.kind != nkSym: continue
+      yield field
+
+template quoeStr(s: string): string = "'" & s & "'"
```
``quoteStr``

Araq commented on compiler/semexprs.nim:
```diff
@@ -2035,25 +2035,238 @@ proc isTupleType(n: PNode): bool =
+    let assignment = locateFieldInInitExpr(field.sym, initExpr)
+    if assignment != nil:
+      if result.len != 0: result.add ", "
+      result.add field.sym.name.s.quoeStr
```
quoteStr

Araq commented on compiler/semexprs.nim:
```diff
@@ -2035,25 +2035,238 @@ proc isTupleType(n: PNode): bool =
+  else:
+    internalAssert false
+
+proc semContuctType(c: PContext, t: PType,
```
semConstructType

Araq commented:
Sorry you need to rebase this. And when you do that, extract code into ``semobjconstr.nim``

zah commented:
Any objections against pushing it straight to devel after rebasing and
addressing the comments?

On Thu, Mar 23, 2017 at 2:21 PM, Andreas Rumpf <notifications@github.com>
wrote:

> Sorry you need to rebase this. And when you do that, extract code into
> semobjconstr.nim
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/nim-lang/Nim/pull/5564#issuecomment-288701707>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAKtZtQ3jUsvyQC3vRTxiVLgIMNbbnHDks5romOugaJpZM4MhgOi>
> .
>

Araq commented:
No. I haven't reviewed it completely but it's localized and I know what it is about. Maybe add a comment about why you take this many passes over the object constructor.

vegansk commented:
https://github.com/vegansk/nimboost is broken after this
